src/v/archival/archival_metadata_stm.cc

```diff
@@ -802,6 +802,26 @@ ss::future<std::error_code> archival_metadata_stm::do_replicate_commands(
 
     const auto current_term = _insync_term;
 
+    // If the caller didn't invoke `sync` before calling `replicate` we
+    // might have some batches which are not applied to the STM yet. These
```

Comment:
Thinking generally about how to make it clearer that this behavior is correct via tests.. it's tough because instantiating the archival stm in isolation and stressing it to exercise these code paths is so tough.

> The only problem that we may have is when the caller invokes 'replicate' without calling 'sync' first. In this case some old batch may trigger the promise and we will get incorrect error code in the 'replicate' method.

Thinking about this specifically, I wonder if there's a way we could artificially/randomly delay the apply fiber and artificially replicate()/sync() from many different fibers in a test. Then the invariant we want to uphold is that all `replicate()` calls that succeeded are reflected in the manifest, and all that failed are not.

I guess it'd be tough to test with the archival stm's current interface because there are so many rules baked into archival operations, but just thinking out loud. It'd be great if this logic were encapsulated in the persisted stm -- then we could implement a dummy stm that's just a kvstore to tests the concurrency mechanics without also having to work around the stm's logical invariants.